{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e8fa5607-b4ed-4c30-b565-29072650242c",
   "metadata": {},
   "source": [
    "# Identify the consititutive heterochromatin states among the mouse genomce by H3K9me3 signals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d041b955-e2e4-40be-8a6a-923a2cde9b6f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages({\n",
    "    library(tidyverse)\n",
    "    library(GenomicRanges)\n",
    "    library(rtracklayer)\n",
    "    library(tmpRpkg)\n",
    "})\n",
    "Sys.setenv(\"_R_USE_PIPEBIND_\" = TRUE)\n",
    "# change the size of figures shown in jupyter\n",
    "options(repr.plot.width=10, repr.plot.height=8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3142b6f3-626c-4dc7-996c-05d3c7d54fbd",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3434 ranges and 1 metadata column:\n",
       "         seqnames            ranges strand | High Signal Region\n",
       "            <Rle>         <IRanges>  <Rle> |        <character>\n",
       "     [1]    chr10   3218900-3276600      * |    Low Mappability\n",
       "     [2]    chr10   3576900-3627700      * |    Low Mappability\n",
       "     [3]    chr10   4191100-4197600      * |    Low Mappability\n",
       "     [4]    chr10   4613500-4615400      * | High Signal Region\n",
       "     [5]    chr10   4761300-4763900      * | High Signal Region\n",
       "     ...      ...               ...    ... .                ...\n",
       "  [3430]     chrY   6530200-6663200      * | High Signal Region\n",
       "  [3431]     chrY   6760200-6835800      * | High Signal Region\n",
       "  [3432]     chrY   6984100-8985400      * | High Signal Region\n",
       "  [3433]     chrY 10638500-41003800      * | High Signal Region\n",
       "  [3434]     chrY 41159200-91744600      * | High Signal Region\n",
       "  -------\n",
       "  seqinfo: 21 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# * meta# * meta\n",
    "projd <- here::here()\n",
    "binSize <- 5000\n",
    "genomeBind <- file.path(projd, \"09.epimem\", \"out\", \"genomeBin\")\n",
    "## load blacklist\n",
    "blfnm <- file.path(projd, \"meta\", \"mm10-blacklist.v2.bed\")\n",
    "blgr <- loadGRfromBed(blfnm, header = TRUE)\n",
    "blgr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7198c44d-1613-450b-9300-5825aaf80101",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 1872052 ranges and 21 metadata columns:\n",
       "            seqnames          ranges strand |   source       type     score\n",
       "               <Rle>       <IRanges>  <Rle> | <factor>   <factor> <numeric>\n",
       "        [1]     chr1 3073253-3074322      + |  HAVANA  gene              NA\n",
       "        [2]     chr1 3073253-3074322      + |  HAVANA  transcript        NA\n",
       "        [3]     chr1 3073253-3074322      + |  HAVANA  exon              NA\n",
       "        [4]     chr1 3102016-3102125      + |  ENSEMBL gene              NA\n",
       "        [5]     chr1 3102016-3102125      + |  ENSEMBL transcript        NA\n",
       "        ...      ...             ...    ... .      ...        ...       ...\n",
       "  [1872048]     chrM     15289-15355      + |  ENSEMBL transcript        NA\n",
       "  [1872049]     chrM     15289-15355      + |  ENSEMBL exon              NA\n",
       "  [1872050]     chrM     15356-15422      - |  ENSEMBL gene              NA\n",
       "  [1872051]     chrM     15356-15422      - |  ENSEMBL transcript        NA\n",
       "  [1872052]     chrM     15356-15422      - |  ENSEMBL exon              NA\n",
       "                phase              gene_id   gene_type     gene_name\n",
       "            <integer>          <character> <character>   <character>\n",
       "        [1]      <NA> ENSMUSG00000102693.1         TEC 4933401J01Rik\n",
       "        [2]      <NA> ENSMUSG00000102693.1         TEC 4933401J01Rik\n",
       "        [3]      <NA> ENSMUSG00000102693.1         TEC 4933401J01Rik\n",
       "        [4]      <NA> ENSMUSG00000064842.1       snRNA       Gm26206\n",
       "        [5]      <NA> ENSMUSG00000064842.1       snRNA       Gm26206\n",
       "        ...       ...                  ...         ...           ...\n",
       "  [1872048]      <NA> ENSMUSG00000064371.1     Mt_tRNA         mt-Tt\n",
       "  [1872049]      <NA> ENSMUSG00000064371.1     Mt_tRNA         mt-Tt\n",
       "  [1872050]      <NA> ENSMUSG00000064372.1     Mt_tRNA         mt-Tp\n",
       "  [1872051]      <NA> ENSMUSG00000064372.1     Mt_tRNA         mt-Tp\n",
       "  [1872052]      <NA> ENSMUSG00000064372.1     Mt_tRNA         mt-Tp\n",
       "                  level      mgi_id          havana_gene        transcript_id\n",
       "            <character> <character>          <character>          <character>\n",
       "        [1]           2 MGI:1918292 OTTMUSG00000049935.1                 <NA>\n",
       "        [2]           2 MGI:1918292 OTTMUSG00000049935.1 ENSMUST00000193812.1\n",
       "        [3]           2 MGI:1918292 OTTMUSG00000049935.1 ENSMUST00000193812.1\n",
       "        [4]           3 MGI:5455983                 <NA>                 <NA>\n",
       "        [5]           3 MGI:5455983                 <NA> ENSMUST00000082908.1\n",
       "        ...         ...         ...                  ...                  ...\n",
       "  [1872048]           3  MGI:102473                 <NA> ENSMUST00000082422.1\n",
       "  [1872049]           3  MGI:102473                 <NA> ENSMUST00000082422.1\n",
       "  [1872050]           3  MGI:102478                 <NA>                 <NA>\n",
       "  [1872051]           3  MGI:102478                 <NA> ENSMUST00000082423.1\n",
       "  [1872052]           3  MGI:102478                 <NA> ENSMUST00000082423.1\n",
       "            transcript_type   transcript_name transcript_support_level\n",
       "                <character>       <character>              <character>\n",
       "        [1]            <NA>              <NA>                     <NA>\n",
       "        [2]             TEC 4933401J01Rik-201                       NA\n",
       "        [3]             TEC 4933401J01Rik-201                       NA\n",
       "        [4]            <NA>              <NA>                     <NA>\n",
       "        [5]           snRNA       Gm26206-201                       NA\n",
       "        ...             ...               ...                      ...\n",
       "  [1872048]         Mt_tRNA         mt-Tt-201                       NA\n",
       "  [1872049]         Mt_tRNA         mt-Tt-201                       NA\n",
       "  [1872050]            <NA>              <NA>                     <NA>\n",
       "  [1872051]         Mt_tRNA         mt-Tp-201                       NA\n",
       "  [1872052]         Mt_tRNA         mt-Tp-201                       NA\n",
       "                    tag    havana_transcript exon_number              exon_id\n",
       "            <character>          <character> <character>          <character>\n",
       "        [1]        <NA>                 <NA>        <NA>                 <NA>\n",
       "        [2]       basic OTTMUST00000127109.1        <NA>                 <NA>\n",
       "        [3]       basic OTTMUST00000127109.1           1 ENSMUSE00001343744.1\n",
       "        [4]        <NA>                 <NA>        <NA>                 <NA>\n",
       "        [5]       basic                 <NA>        <NA>                 <NA>\n",
       "        ...         ...                  ...         ...                  ...\n",
       "  [1872048]       basic                 <NA>        <NA>                 <NA>\n",
       "  [1872049]       basic                 <NA>           1 ENSMUSE00000521549.1\n",
       "  [1872050]        <NA>                 <NA>        <NA>                 <NA>\n",
       "  [1872051]       basic                 <NA>        <NA>                 <NA>\n",
       "  [1872052]       basic                 <NA>           1 ENSMUSE00000521550.1\n",
       "             protein_id      ccdsid         ont\n",
       "            <character> <character> <character>\n",
       "        [1]        <NA>        <NA>        <NA>\n",
       "        [2]        <NA>        <NA>        <NA>\n",
       "        [3]        <NA>        <NA>        <NA>\n",
       "        [4]        <NA>        <NA>        <NA>\n",
       "        [5]        <NA>        <NA>        <NA>\n",
       "        ...         ...         ...         ...\n",
       "  [1872048]        <NA>        <NA>        <NA>\n",
       "  [1872049]        <NA>        <NA>        <NA>\n",
       "  [1872050]        <NA>        <NA>        <NA>\n",
       "  [1872051]        <NA>        <NA>        <NA>\n",
       "  [1872052]        <NA>        <NA>        <NA>\n",
       "  -------\n",
       "  seqinfo: 22 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "          3prime_overlapping_ncRNA                          antisense \n",
       "                                16                              19316 \n",
       "     bidirectional_promoter_lncRNA                          IG_C_gene \n",
       "                              1211                                246 \n",
       "                   IG_C_pseudogene                          IG_D_gene \n",
       "                                 5                                 78 \n",
       "                   IG_D_pseudogene                          IG_J_gene \n",
       "                                 9                                 64 \n",
       "                        IG_LV_gene                      IG_pseudogene \n",
       "                                22                                  7 \n",
       "                         IG_V_gene                    IG_V_pseudogene \n",
       "                              2000                                587 \n",
       "                           lincRNA                       macro_lncRNA \n",
       "                             40389                                 10 \n",
       "                             miRNA                           misc_RNA \n",
       "                              6606                               1686 \n",
       "                           Mt_rRNA                            Mt_tRNA \n",
       "                                 6                                 66 \n",
       "            polymorphic_pseudogene               processed_pseudogene \n",
       "                              2099                              31103 \n",
       "              processed_transcript                     protein_coding \n",
       "                             11438                            1711899 \n",
       "                        pseudogene                           ribozyme \n",
       "                               271                                 66 \n",
       "                              rRNA                             scaRNA \n",
       "                              1062                                153 \n",
       "                             scRNA                     sense_intronic \n",
       "                                 3                               1305 \n",
       "                 sense_overlapping                             snoRNA \n",
       "                               174                               4521 \n",
       "                             snRNA                               sRNA \n",
       "                              4149                                  6 \n",
       "                               TEC                          TR_C_gene \n",
       "                              9899                                115 \n",
       "                         TR_D_gene                          TR_J_gene \n",
       "                                18                                292 \n",
       "                   TR_J_pseudogene                          TR_V_gene \n",
       "                                30                               1350 \n",
       "                   TR_V_pseudogene   transcribed_processed_pseudogene \n",
       "                               115                               2079 \n",
       "    transcribed_unitary_pseudogene transcribed_unprocessed_pseudogene \n",
       "                               659                               4329 \n",
       " translated_unprocessed_pseudogene                 unitary_pseudogene \n",
       "                                12                                419 \n",
       "            unprocessed_pseudogene \n",
       "                             12162 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load gtf\n",
    "mm10 <- import(file.path(projd, \"meta\", \"gencode.vM25.annotation.gtf\"))\n",
    "mm10\n",
    "table(mm10$gene_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1902585d-94c5-4a98-8d22-6355229b918f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## get TE annotations in the genome\n",
    "loadRepeatMasker <- function() {\n",
    "    colnms <- c(\"bin\", \"swScore\", \"milliDiv\", \"milliDel\", \"milliIns\", \"genoName\", \"genoStart\", \"genoEnd\", \n",
    "               \"genoLeft\", \"strand\", \"repName\", \"repClass\", \"repFamliy\", \"repStart\", \"repEnd\", \"repLeft\", \n",
    "               \"id\")\n",
    "    r <- data.table::fread(\n",
    "        file = file.path(projd, \"meta\", \"mm10.RepeatMasker.txt\"), \n",
    "        sep = \"\\t\", header = FALSE) |>\n",
    "      setNames(object = _, nm = colnms)\n",
    "    return(r)\n",
    "}\n",
    "rawTEs <- loadRepeatMasker()\n",
    "## other heterochromatin annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fc3b85c4-14fa-4ea2-b024-b55a0e5d47d8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 17</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>bin</th><th scope=col>swScore</th><th scope=col>milliDiv</th><th scope=col>milliDel</th><th scope=col>milliIns</th><th scope=col>genoName</th><th scope=col>genoStart</th><th scope=col>genoEnd</th><th scope=col>genoLeft</th><th scope=col>strand</th><th scope=col>repName</th><th scope=col>repClass</th><th scope=col>repFamliy</th><th scope=col>repStart</th><th scope=col>repEnd</th><th scope=col>repLeft</th><th scope=col>id</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>585</td><td> 463</td><td> 13</td><td>  6</td><td>17</td><td>chr1</td><td>10000</td><td>10468</td><td>-248945954</td><td>+</td><td>(TAACCC)n</td><td>Simple_repeat</td><td>Simple_repeat</td><td>    1</td><td> 471</td><td>  0</td><td>1</td></tr>\n",
       "\t<tr><td>585</td><td>3612</td><td>114</td><td>215</td><td>13</td><td>chr1</td><td>10468</td><td>11447</td><td>-248944975</td><td>-</td><td>TAR1     </td><td>Satellite    </td><td>telo         </td><td> -399</td><td>1712</td><td>483</td><td>2</td></tr>\n",
       "\t<tr><td>585</td><td> 484</td><td>251</td><td>132</td><td> 0</td><td>chr1</td><td>11504</td><td>11675</td><td>-248944747</td><td>-</td><td>L1MC5a   </td><td>LINE         </td><td>L1           </td><td>-2382</td><td> 395</td><td>199</td><td>3</td></tr>\n",
       "\t<tr><td>585</td><td> 239</td><td>294</td><td> 19</td><td>10</td><td>chr1</td><td>11677</td><td>11780</td><td>-248944642</td><td>-</td><td>MER5B    </td><td>DNA          </td><td>hAT-Charlie  </td><td>  -74</td><td> 104</td><td>  1</td><td>4</td></tr>\n",
       "\t<tr><td>585</td><td> 318</td><td>230</td><td> 37</td><td> 0</td><td>chr1</td><td>15264</td><td>15355</td><td>-248941067</td><td>-</td><td>MIR3     </td><td>SINE         </td><td>MIR          </td><td> -119</td><td> 143</td><td> 49</td><td>5</td></tr>\n",
       "\t<tr><td>585</td><td>  18</td><td>232</td><td>  0</td><td>19</td><td>chr1</td><td>15797</td><td>15849</td><td>-248940573</td><td>+</td><td>(TGCTCC)n</td><td>Simple_repeat</td><td>Simple_repeat</td><td>    1</td><td>  52</td><td>  0</td><td>6</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 17\n",
       "\\begin{tabular}{lllllllllllllllll}\n",
       " bin & swScore & milliDiv & milliDel & milliIns & genoName & genoStart & genoEnd & genoLeft & strand & repName & repClass & repFamliy & repStart & repEnd & repLeft & id\\\\\n",
       " <int> & <int> & <int> & <int> & <int> & <chr> & <int> & <int> & <int> & <chr> & <chr> & <chr> & <chr> & <int> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t 585 &  463 &  13 &   6 & 17 & chr1 & 10000 & 10468 & -248945954 & + & (TAACCC)n & Simple\\_repeat & Simple\\_repeat &     1 &  471 &   0 & 1\\\\\n",
       "\t 585 & 3612 & 114 & 215 & 13 & chr1 & 10468 & 11447 & -248944975 & - & TAR1      & Satellite     & telo          &  -399 & 1712 & 483 & 2\\\\\n",
       "\t 585 &  484 & 251 & 132 &  0 & chr1 & 11504 & 11675 & -248944747 & - & L1MC5a    & LINE          & L1            & -2382 &  395 & 199 & 3\\\\\n",
       "\t 585 &  239 & 294 &  19 & 10 & chr1 & 11677 & 11780 & -248944642 & - & MER5B     & DNA           & hAT-Charlie   &   -74 &  104 &   1 & 4\\\\\n",
       "\t 585 &  318 & 230 &  37 &  0 & chr1 & 15264 & 15355 & -248941067 & - & MIR3      & SINE          & MIR           &  -119 &  143 &  49 & 5\\\\\n",
       "\t 585 &   18 & 232 &   0 & 19 & chr1 & 15797 & 15849 & -248940573 & + & (TGCTCC)n & Simple\\_repeat & Simple\\_repeat &     1 &   52 &   0 & 6\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 17\n",
       "\n",
       "| bin &lt;int&gt; | swScore &lt;int&gt; | milliDiv &lt;int&gt; | milliDel &lt;int&gt; | milliIns &lt;int&gt; | genoName &lt;chr&gt; | genoStart &lt;int&gt; | genoEnd &lt;int&gt; | genoLeft &lt;int&gt; | strand &lt;chr&gt; | repName &lt;chr&gt; | repClass &lt;chr&gt; | repFamliy &lt;chr&gt; | repStart &lt;int&gt; | repEnd &lt;int&gt; | repLeft &lt;int&gt; | id &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 585 |  463 |  13 |   6 | 17 | chr1 | 10000 | 10468 | -248945954 | + | (TAACCC)n | Simple_repeat | Simple_repeat |     1 |  471 |   0 | 1 |\n",
       "| 585 | 3612 | 114 | 215 | 13 | chr1 | 10468 | 11447 | -248944975 | - | TAR1      | Satellite     | telo          |  -399 | 1712 | 483 | 2 |\n",
       "| 585 |  484 | 251 | 132 |  0 | chr1 | 11504 | 11675 | -248944747 | - | L1MC5a    | LINE          | L1            | -2382 |  395 | 199 | 3 |\n",
       "| 585 |  239 | 294 |  19 | 10 | chr1 | 11677 | 11780 | -248944642 | - | MER5B     | DNA           | hAT-Charlie   |   -74 |  104 |   1 | 4 |\n",
       "| 585 |  318 | 230 |  37 |  0 | chr1 | 15264 | 15355 | -248941067 | - | MIR3      | SINE          | MIR           |  -119 |  143 |  49 | 5 |\n",
       "| 585 |   18 | 232 |   0 | 19 | chr1 | 15797 | 15849 | -248940573 | + | (TGCTCC)n | Simple_repeat | Simple_repeat |     1 |   52 |   0 | 6 |\n",
       "\n"
      ],
      "text/plain": [
       "  bin swScore milliDiv milliDel milliIns genoName genoStart genoEnd genoLeft  \n",
       "1 585  463     13        6      17       chr1     10000     10468   -248945954\n",
       "2 585 3612    114      215      13       chr1     10468     11447   -248944975\n",
       "3 585  484    251      132       0       chr1     11504     11675   -248944747\n",
       "4 585  239    294       19      10       chr1     11677     11780   -248944642\n",
       "5 585  318    230       37       0       chr1     15264     15355   -248941067\n",
       "6 585   18    232        0      19       chr1     15797     15849   -248940573\n",
       "  strand repName   repClass      repFamliy     repStart repEnd repLeft id\n",
       "1 +      (TAACCC)n Simple_repeat Simple_repeat     1     471     0     1 \n",
       "2 -      TAR1      Satellite     telo           -399    1712   483     2 \n",
       "3 -      L1MC5a    LINE          L1            -2382     395   199     3 \n",
       "4 -      MER5B     DNA           hAT-Charlie     -74     104     1     4 \n",
       "5 -      MIR3      SINE          MIR            -119     143    49     5 \n",
       "6 +      (TGCTCC)n Simple_repeat Simple_repeat     1      52     0     6 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(rawTEs)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "b46d83ba-57b2-4550-98f1-7a72dcee8bb4",
   "metadata": {},
   "source": [
    "# 1. Calculate the H3K9me3 signals in 5kb-size genome bins\n",
    "# 2. Filter black list and sex chromosomes\n",
    "# 2. Identify the regions that are show high signals, medium ones, and low signals\n",
    "     - The regions should have low H3K27ac, H3K27me3, H3K4me1 signals (at least lower than Q50).\n",
    "     - gene body +- 5kb regions should be excluded\n",
    "# 3. Overlap with \n",
    "     - the annotated TE regions or other known constitutive heterochromatin (positive control)\n",
    "     - promoter regions, ATAC-seq peak regions we have identified (negative control)\n",
    "     - cross different cell types, are they conservative? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bd9498c6-258b-4d0a-a8c9-98e892396348",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# * functions\n",
    "histbwd <- file.path(projd, \"09.epimem\", \"out\", \"genomeBin\")\n",
    "loadAllHistoneSignals <- function(sc) {\n",
    "  r1 <- read.csv(file.path(histbwd, paste0(sc, \".H3K9me3.5000.csv\")), sep = \",\", header = TRUE)\n",
    "  rownames(r1) <- r1$region\n",
    "  r2 <- read.csv(file.path(histbwd, paste0(sc, \".H3K27ac_H3K27me3_H3K4me1.5000.csv\")), sep = \",\", header = TRUE)\n",
    "  rownames(r2) <- r2$region\n",
    "  r <- merge(r1, r2, by = \"region\", all = FALSE)\n",
    "  return(r)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "86f826f8-abbb-40ca-9d4f-1e9e320953ec",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>region</th><th scope=col>H3K9me3</th><th scope=col>H3K27ac</th><th scope=col>H3K27me3</th><th scope=col>H3K4me1</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chr1:1-5000             </td><td>0.0000000</td><td>0.0000000</td><td>0.0000000</td><td>0.0000000</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chr1:100000001-100005000</td><td>1.3969261</td><td>0.3557545</td><td>0.9780488</td><td>0.3610463</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chr1:10000001-10005000  </td><td>0.1817366</td><td>0.3313600</td><td>1.1783989</td><td>0.2195246</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chr1:1000001-1005000    </td><td>0.0000000</td><td>0.0000000</td><td>0.0000000</td><td>0.0000000</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chr1:100001-105000      </td><td>0.0000000</td><td>0.0000000</td><td>0.0000000</td><td>0.0000000</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chr1:100005001-100010000</td><td>0.6263849</td><td>0.2784645</td><td>0.7742647</td><td>0.2754316</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & region & H3K9me3 & H3K27ac & H3K27me3 & H3K4me1\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & chr1:1-5000              & 0.0000000 & 0.0000000 & 0.0000000 & 0.0000000\\\\\n",
       "\t2 & chr1:100000001-100005000 & 1.3969261 & 0.3557545 & 0.9780488 & 0.3610463\\\\\n",
       "\t3 & chr1:10000001-10005000   & 0.1817366 & 0.3313600 & 1.1783989 & 0.2195246\\\\\n",
       "\t4 & chr1:1000001-1005000     & 0.0000000 & 0.0000000 & 0.0000000 & 0.0000000\\\\\n",
       "\t5 & chr1:100001-105000       & 0.0000000 & 0.0000000 & 0.0000000 & 0.0000000\\\\\n",
       "\t6 & chr1:100005001-100010000 & 0.6263849 & 0.2784645 & 0.7742647 & 0.2754316\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 5\n",
       "\n",
       "| <!--/--> | region &lt;chr&gt; | H3K9me3 &lt;dbl&gt; | H3K27ac &lt;dbl&gt; | H3K27me3 &lt;dbl&gt; | H3K4me1 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 1 | chr1:1-5000              | 0.0000000 | 0.0000000 | 0.0000000 | 0.0000000 |\n",
       "| 2 | chr1:100000001-100005000 | 1.3969261 | 0.3557545 | 0.9780488 | 0.3610463 |\n",
       "| 3 | chr1:10000001-10005000   | 0.1817366 | 0.3313600 | 1.1783989 | 0.2195246 |\n",
       "| 4 | chr1:1000001-1005000     | 0.0000000 | 0.0000000 | 0.0000000 | 0.0000000 |\n",
       "| 5 | chr1:100001-105000       | 0.0000000 | 0.0000000 | 0.0000000 | 0.0000000 |\n",
       "| 6 | chr1:100005001-100010000 | 0.6263849 | 0.2784645 | 0.7742647 | 0.2754316 |\n",
       "\n"
      ],
      "text/plain": [
       "  region                   H3K9me3   H3K27ac   H3K27me3  H3K4me1  \n",
       "1 chr1:1-5000              0.0000000 0.0000000 0.0000000 0.0000000\n",
       "2 chr1:100000001-100005000 1.3969261 0.3557545 0.9780488 0.3610463\n",
       "3 chr1:10000001-10005000   0.1817366 0.3313600 1.1783989 0.2195246\n",
       "4 chr1:1000001-1005000     0.0000000 0.0000000 0.0000000 0.0000000\n",
       "5 chr1:100001-105000       0.0000000 0.0000000 0.0000000 0.0000000\n",
       "6 chr1:100005001-100010000 0.6263849 0.2784645 0.7742647 0.2754316"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "90.9749137480084"
      ],
      "text/latex": [
       "90.9749137480084"
      ],
      "text/markdown": [
       "90.9749137480084"
      ],
      "text/plain": [
       "[1] 90.97491"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>10%</dt><dd>0.187499155316097</dd><dt>25%</dt><dd>0.307502731673977</dd><dt>50%</dt><dd>0.509504905848008</dd><dt>75%</dt><dd>0.872021068633497</dd><dt>90%</dt><dd>1.5129094216556</dd><dt>95%</dt><dd>2.04684743617243</dd><dt>99%</dt><dd>3.56141279281563</dd><dt>99.9%</dt><dd>5.80354157739084</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[10\\textbackslash{}\\%] 0.187499155316097\n",
       "\\item[25\\textbackslash{}\\%] 0.307502731673977\n",
       "\\item[50\\textbackslash{}\\%] 0.509504905848008\n",
       "\\item[75\\textbackslash{}\\%] 0.872021068633497\n",
       "\\item[90\\textbackslash{}\\%] 1.5129094216556\n",
       "\\item[95\\textbackslash{}\\%] 2.04684743617243\n",
       "\\item[99\\textbackslash{}\\%] 3.56141279281563\n",
       "\\item[99.9\\textbackslash{}\\%] 5.80354157739084\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "10%\n",
       ":   0.18749915531609725%\n",
       ":   0.30750273167397750%\n",
       ":   0.50950490584800875%\n",
       ":   0.87202106863349790%\n",
       ":   1.512909421655695%\n",
       ":   2.0468474361724399%\n",
       ":   3.5614127928156399.9%\n",
       ":   5.80354157739084\n",
       "\n"
      ],
      "text/plain": [
       "      10%       25%       50%       75%       90%       95%       99%     99.9% \n",
       "0.1874992 0.3075027 0.5095049 0.8720211 1.5129094 2.0468474 3.5614128 5.8035416 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# analyze one k9 as example\n",
    "sc <- \"326_OPC_NN\"\n",
    "k9 <- loadAllHistoneSignals(sc)\n",
    "head(k9)\n",
    "max(k9$H3K9me3)\n",
    "quantile(k9$H3K9me3, probs = c(0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99, 0.999))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3e298d0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6580411-fafa-41ae-9a92-2c18f5b224bc",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f182668-66b8-4ab9-b234-f5a3628172a3",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acd17817-10be-40df-9ba6-5578f93c2da1",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# remove blacklist and sex chromosomes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da472b41-ad4b-4b6f-a622-b26040fa9b72",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# The regions should have low H3K27ac, H3K27me3, H3K4me1 signals (at least lower than Q50)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
